hazelcast/src/main/java/com/hazelcast/map/impl/MapMergeRunnable.java

```diff
@@ -0,0 +1,127 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.map.impl;
+
+import com.hazelcast.config.InMemoryFormat;
+import com.hazelcast.config.MapConfig;
+import com.hazelcast.config.MergePolicyConfig;
+import com.hazelcast.core.EntryView;
+import com.hazelcast.map.impl.operation.MapOperationProvider;
+import com.hazelcast.map.impl.record.Record;
+import com.hazelcast.map.impl.recordstore.RecordStore;
+import com.hazelcast.map.merge.MapMergePolicy;
+import com.hazelcast.nio.serialization.Data;
+import com.hazelcast.spi.Operation;
+import com.hazelcast.spi.OperationFactory;
+import com.hazelcast.spi.SplitBrainMergePolicy;
+import com.hazelcast.spi.impl.AbstractMergeRunnable;
+import com.hazelcast.spi.merge.MergingEntryHolder;
+import com.hazelcast.util.Clock;
+import com.hazelcast.util.function.BiConsumer;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import static com.hazelcast.map.impl.EntryViews.createSimpleEntryView;
+import static com.hazelcast.spi.impl.merge.MergingHolders.createMergeHolder;
+
+class MapMergeRunnable extends AbstractMergeRunnable<RecordStore, MergingEntryHolder<Data, Data>> {
+
+    private final MapServiceContext mapServiceContext;
+    private final MapSplitBrainHandlerService mapSplitBrainHandlerService;
+
+    MapMergeRunnable(Map<String, Collection<RecordStore>> collectedStores,
+                     Map<String, Collection<RecordStore>> collectedStoresWithLegacyPolicies,
+                     Collection<RecordStore> backupStores, MapServiceContext mapServiceContext,
+                     MapSplitBrainHandlerService mapSplitBrainHandlerService) {
+        super(MapService.SERVICE_NAME, collectedStores, collectedStoresWithLegacyPolicies,
+                backupStores, mapServiceContext.getNodeEngine());
+
+        this.mapServiceContext = mapServiceContext;
+        this.mapSplitBrainHandlerService = mapSplitBrainHandlerService;
+    }
+
+    @Override
+    protected void consumeStore(RecordStore store, BiConsumer<Integer, MergingEntryHolder<Data, Data>> consumer) {
+        long now = Clock.currentTimeMillis();
+        int partitionId = store.getPartitionId();
+
+        Iterator<Record> iterator = store.iterator(now, false);
+        while (iterator.hasNext()) {
+            Record record = iterator.next();
+
+            Data dataValue = toData(record.getValue());
```

Comment:
if data is unmodifiable, no problem.

Reply:
Do you mean that the HD data can be accessed concurrently if it is immutable or can it be accessed by only one thread at a time?

Reply:
- (Pooled) HD memory can be allocated/written/disposed only and only by partition thread itself. Other threads must not try to allocate/write/dispose a native memory block that's belonging to a specific partition thread.
- (Pooled) HD memory can be *read* by any thread (even concurrently) _if and only if_ it's guaranteed that native memory content is not going to be overwritten and native memory block is not going to be disposed. Otherwise, readers must be exclusively synchronized with writer (partition thread), this is what we do for Hot Restart integration with HD IMap/ICache.  

As far as I understand, in this case, record stores are read by non-partition threads but they are not accessible by partition threads anymore. And they are disposed by partition threads after merge.

Reply:
Got it, thanks Mehmet! I was always unclear what do we actually mean by that catch-phrase "HD memory can only be accessed from partition thread".